check.go

```diff
@@ -0,0 +1,1086 @@
+							}
+							err = fmt.Errorf("container %s: data item %q: %w", id, key, err)
+							report.Containers[id] = append(report.Containers[id], err)
+							return
+						}
+						if int64(len(data)) != container.BigDataSizes[key] {
+							err = fmt.Errorf("container %s: data item %q: %w", id, key, ErrContainerDataIncorrectSize)
+							report.Containers[id] = append(report.Containers[id], err)
+							return
+						}
+					}()
+				}
+			}
+			// Look at the container's base image.  If the image has errors, the image's errors
+			// are the container's errors.
+			if container.ImageID != "" {
+				if _, checked := examinedImages[container.ImageID]; !checked {
+					err := fmt.Errorf("image %s: %w", container.ImageID, ErrContainerImageMissing)
+					report.Containers[id] = append(report.Containers[id], err)
+				}
+				if len(report.Images[container.ImageID]) > 0 {
+					report.Containers[id] = append(report.Containers[id], report.Images[container.ImageID]...)
+				}
+				if len(report.ROImages[container.ImageID]) > 0 {
+					report.Containers[id] = append(report.Containers[id], report.ROImages[container.ImageID]...)
+				}
+			}
+			// Count the container's layer as referenced.
+			if container.LayerID != "" {
+				referencedLayers[container.LayerID] = true
+			}
+		}
+		return false, nil
+	}); done {
+		return CheckReport{}, err
+	}
+
+	// Now go back through all of the layer stores, and flag any layers which don't belong
+	// to an image or a container, and has been around longer than we can reasonably expect
+	// such a layer to be present before a corresponding image record is added.
+	if errored, err := s.readAllLayerStores(func(store roLayerStore) (bool, error) {
+		if isReadWrite := roLayerStoreIsReallyReadWrite(store); !isReadWrite {
+			return false, nil
+		}
+		layers, err := store.Layers()
+		if err != nil {
+			return true, err
+		}
+		for _, layer := range layers {
+			maximumAge := defaultMaximumUnreferencedLayerAge
+			if options.LayerUnreferencedMaximumAge != nil {
+				maximumAge = *options.LayerUnreferencedMaximumAge
+			}
+			if referenced := referencedLayers[layer.ID]; !referenced {
+				if layer.Created.IsZero() || layer.Created.Add(maximumAge).Before(time.Now()) {
+					// Either we don't (and never will) know when this layer was
+					// created, or it was created far enough in the past that we're
+					// reasonably sure it's not part of an image that's being written
+					// right now.
+					err := fmt.Errorf("layer %s: %w", layer.ID, ErrLayerUnreferenced)
+					report.Layers[layer.ID] = append(report.Layers[layer.ID], err)
+				}
+			}
+		}
+		return false, nil
+	}); errored {
+		return CheckReport{}, err
+	}
+
+	// If the driver can tell us about which layers it knows about, we should have previously
+	// examined all of them.  Any that we didn't are probably just wasted space.
+	// Note: if the driver doesn't support enumerating layers, it returns ErrNotSupported.
+	if err := s.startUsingGraphDriver(); err != nil {
```

Comment:
Hum, should the graph lock be held for the lifetime of this function, to ensure a consistent view?